lib/Blin/Tester/Zef.rakumod

```diff
@@ -54,9 +54,6 @@ submethod TWEAK ( ) {
 }
 
 method test-command( ::?CLASS:D: :$testable!, :$install-path!, :$module-name! ) {
-    $*EXECUTABLE.absolute,
-    ‘-I’,
-    $!path,
```

Comment:
I don't doubt removing this somehow works around the issue. However, it speaks to a bigger issue somewhere. We know zef does not get installed, and that `$!binary = $!path.add: 'bin/zef';`, so what e.g. `Zef::CLI`, `Zef::Repository`, etc libraries could be getting used to run that `$!binary`? It would seem like it has to be seeing either some installed version of zef (which isn't going to be the same as what the cloning and whatnot intends to supply), or there is a `PATH` containing those libs still. Or maybe `$*EXECUTABLE` isn't getting set right (I dunno if Blin does any tricks for bisecting that might muck that up, like swapping the shebang or using some wrapper/shim).

Unfortunately I don't have a good suggestion on how to proceed. It would probably be worth it to print out the command that was getting run (to see what `$*EXECUTABLE.absolute` `$!path` etc are) and then run it manually to see if you get the same error. Seeing the command being used might be enough for me to figure out what is going on as well.

Reply:
The resulting test-command is (via `say .raku` formatting):

>["/home/koto/.rakudobrew/versions/moar-2020.12/install/bin/rakudo-m", "-I", IO::Path.new("data/zef", :SPEC(IO::Spec::Unix), :CWD("/home/koto/Devel/Blin")), IO::Path.new("data/zef/bin/zef", :SPEC(IO::Spec::Unix), :CWD("/home/koto/Devel/Blin")), "--config-path=data/zef-config.json", ("--verbose", "--force-build", "--force-install"), "--force-test", ("--/depends", "--/test-depends", "--/build-depends"), "install", "--to=inst#/tmp/_BVi5L7XuE", "Cro::Core"]

(I am using rakudobrew raku executable here, so $*EXECUTABLE.absolute is technically correct). The data/zef/bin/zef is present and contains:

```
!/usr/bin/env perl6

use Zef::CLI;
```

The error arises from nqp land (https://github.com/Raku/nqp/blob/master/src/HLL/Compiler.nqp#L437) and without rebuilding it will be hard to dig further.

Reply:
When run as the "same" command using shell, `/home/koto/.rakudobrew/versions/moar-2020.12/install/bin/rakudo-m -Idata/zef data/zef/bin/zef --config-path=data/zef-config.json --verbose --force-build --force-install --force-test --/depends --/test-depends --/build-depends install --to=inst#/tmp/rMf7bWRnkm Cro::Core`, there are no issues.

Reply:
Also, running a Blin with zef now takes twice the time, 16 hours vs ~8 hours. I am a bit doubtful simply adding a new eco sources could be the culprit for such timings, but I wonder why nobody notices 2x slowdown (or somewhere near, but still a regression) in scripts if it's a Rakudo thing recently introduced.

Reply:
FYI: We just lost power again so I might not be able to investigate this for a day or two

Reply:
Just some wild guesses from looking at this on my phone:

1. various parts of the command in the `.raku` output are grouped but shouldn't be like with `"--config-path=data/zef-config.json", ("--verbose", "--force-build", "--force-install")` which should probably be `"--config-path=data/zef-config.json", "--verbose", "--force-build", "--force-install"`
2. most of the paths shown in the `.raku` output are relative `IO::Path` objects, when maybe they should all have `.absolute` called on them (so its not an `IO::Path` getting passed to `spawn`, and so the cwd changing for whatever reason won't affect the command)

Reply:
> Also, running a Blin with zef now takes twice the time, 16 hours vs ~8 hours.

Yeah that shouldn't be the case, however -- ecosystems are searched in parallel so adding a new one should not slow down a powerful machine unless the network of the new ecosystem is significantly slower (particularly wherever the `.git` or `.tar` file comes from). I wouldn't expect this to be such a bottle neck, but its possible.

Another thing might be all the typing I added a few months ago. Remember we did a Blin run after those changes, but its possible that ran an old version of zef (since that seemed to be the case a month or two later) and thus the timings didn't appear to change. However I did not notice such a performance regressing in my own testing, so I'm not sure how likely this really is... I certainly wouldn't expect it to take twice a long.

If there are timings available to compare I'd want to investigate which phase (fetch, extract, test, install) is taking significantly more time for any given distribution. For example if its the search phase then its probably typing or ecosystem searching related, if its fetching its probably the new ecosystem content server is slow (doubtful), if its everything then its potentially those zef changes or a raku performance regression.

Reply:
Two more things:

1. it could be unnecessary precomp due to those those relative paths I mentioned earlier, since if the cwd changes (and thus a relative `-I $!path` changes) it would re-precompile `$!binary`
2. it could be `PERL6LIB` related since zef changed to using `RAKULIB` internally (this should be easy to test though... just set `RAKULIB` anywhere in Blin it sets `PERL6LIB`)

Reply:
> The error arises from nqp land (https://github.com/Raku/nqp/blob/master/src/HLL/Compiler.nqp#L437)

Ah, its unfortunate it doesn't output `$filename` (if it exists) in the error... maybe it should. If we knew what file it was trying to open it might be obvious what is happening. A naive example could be some precomp getting read for some weird reason:

```
$ raku -e 'my $in-handle := open(".precomp/C055EB2A59E5EF8913C7028EEFA9C0658BDA5D3A/78/78667563C122DEC985FF889843297317FDAED4F8", :r, :enc("utf8")); say $in-handle.slurp'

Malformed UTF-8 near bytes 00 00 88
  in block <unit> at -e line 1
 ```

Reply:
I think one mystery is solved. The test command is executed using rakudo version for certain commit (as it bisects, so runs with different versions).
https://github.com/Raku/Blin/blob/master/lib/Blin/Processing.rakumod#L399-L401 <- here rakudo binary is already passed ($binary-path looks like e.g. `/tmp/whateverable/rakudo-moar/45b4e6464d5437526aac6fa0c0822c2c90c0cb2c/bin/perl6`) and if you add it to the `test-command` itself as was done, it calls something like `raku raku zef` instead of just `raku zef`, which then fails, that's why removing it "fixes" the issue.

Reply:
Ah yes. And the `-I.` is otherwise handled by https://github.com/Raku/Blin/blob/99e3335fa955db0e44cc86601882ca8302eabd26/lib/Blin/Processing.rakumod#L379
(although passing it in via `-I` would be better, since doing it the current way effectively treats the tester as already installed i.e. modules could `use Zef` and not fail when they should)

And speaking of `PERL6LIB`, around that spot above it might speed things up to add `%tweaked-env<RAKULIB> := %tweaked-env<PERL6LIB>`. If it does then that is a good change to have anyway, but I could also make a change in zef to fix it for anyone else in the mean time.

Reply:
Patched. Is there an easy way to get installation timings to check what can slow down zef?

Reply:
> Patched. Is there an easy way to get installation timings to check what can slow down zef?

To get the timings for each step requires each line of output to have a timestamp prefix -- I think Blin already has this, but most ci we use also does this for instance. If I had the output of a before and after that included those timestamps I can look to see which parts of the output are taking longer to be printed. Thus if I look at how long it takes to find the dependencies of a given distribution and its the same before/after, but the tests take longer (and this pattern is the same for all modules) we can get a good idea of where to look.